### PR TITLE
Powershell Guidelines

### PR DESCRIPTION
**Pull request by jaredpar**
Powershell is used pretty heavily throughout our repo.  This establishes
some guidelines to approach authoring these scripts.

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-infrastructure

File: build/scripts/cibuild.ps1 [hunk 1]
```diff
@@ -45,6 +45,12 @@ function Terminate-BuildProcesses() {
 
 try {
     . (Join-Path $PSScriptRoot "build-utils.ps1")
+
+    if ($badArgs -ne $null) {
```
Comment (sharwell) [inline build/scripts/cibuild.ps1]: ❗️ I believe this needs to come after the `Push-Location` line, since the `finally` will execute.

Comment (jaredpar) [inline build/scripts/cibuild.ps1]: Indeed ...

File: docs/contributing/Powershell Guidelines.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        Print-Usage 
+        exit 1
+    }
+}
+
+```
+
+## Execution
+
+Invoking windows commands should be done via the Exec function.  This adds automatic error detection
```
Comment (tannergooding) [inline docs/contributing/Powershell Guidelines.md]: You should probably add a note that this only works for executables that guarantee a return value of not-zero is a failure. This wrapping will not work for something like `robocopy` (just an example, `robocopy` should never be called from powershell anyways), which returns values other than 0 to represent success.

File: docs/contributing/Powershell Guidelines.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+Powershell is primarily used in this repo as the backbone of our infrastructure.  As such 
+Powershell scripts need to be portable, reliable and take advantage of stop on error 
+approaches.  The guidelines below are meant to push scripts to this mindset of execution.
+
+# Coding Style
+
+1. Opening braces always go at the end of an expression / statement. 
+1. Closing braces always occur on an otherwise empty line.
+1. We use four space indentation.
```
Comment (tannergooding) [inline docs/contributing/Powershell Guidelines.md]: I realize this is a matter of opinion, but it is fairly standard practice for shell scripts to use 2 space indentation.

Comment (jaredpar) [inline docs/contributing/Powershell Guidelines.md]: All the other repos I looked at used 4 space indent for powershell.

File: docs/contributing/Powershell Guidelines.md (hunk 2 above)
Comment (jaredpar) [inline docs/contributing/Powershell Guidelines.md]: Yeah ... those types of programs are just evil

File: docs/contributing/Powershell Guidelines.md (hunk 3 above)
Comment (tannergooding) [inline docs/contributing/Powershell Guidelines.md]: Fair enough. I'm probably just used to the convention for *.sh and *.cmd/bat

**Pull request merged**